File: manual/linear_model.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) {
  stop("Usage: Rscript linear_regression_r.R <filename> <x_column> <y_column>")
}

filename <- args[1]
x_col <- args[2]
y_col <- args[3]

dataset <- read.csv(filename)
formula <- as.formula(paste(y_col, "~", x_col))
model <- lm(formula, data=dataset)

slope <- coef(model)[2]
intercept <- coef(model)[1]
pred <- predict(model)
r <- cor(dataset[[x_col]], dataset[[y_col]])
mse <- mean((dataset[[y_col]] - pred)^2)
cat("Slope:", slope, "\n")
cat("Intercept:", intercept, "\n")
cat("Correlation coefficient:", r, "\n")
cat("Mean Squared Error:", mse, "\n")

library(ggplot2)
plot <- ggplot(dataset,aes(x = .data[[x_col]], y = .data[[y_col]])) +
  geom_point(colour = 'blue') +
  geom_smooth(method = "lm",formula = y ~ x,colour = 'green',se = FALSE) +
  annotate("text", x = 1.5,y = max(dataset[[y_col]]) - 10000,label = paste("y =", round(slope, 2), "x +", round(intercept, 2),"\nr =", round(r,      2),"\nMSE =", round(mse, 2))) +
  ggtitle(paste(y_col,"vs", x_col)) +
  xlab(x_col) +
  ylab(y_col)

ggsave("regression_plot_r.png",plot,width = 7,height = 7, units = "in")

File: manual/linear_model.py
#!/usr/bin/env python
# coding: utf-8

# This notebook demonstrates a simple linear regression analysis using python to model Salary based on Years of Experience.

import sys
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error

if len(sys.argv) != 4:
    print("Usage: python linear_model.py <filename> <x_column> <y_column>")
    sys.exit(1)

filename = sys.argv[1]
x_title = sys.argv[2]
y_title = sys.argv[3]

dataset = pd.read_csv(filename)
x = dataset[x_title]
y = dataset[y_title]

# Linear regression
slope, intercept, r_value, p_value, std_err = linregress(x, y)
y_pred = slope * x + intercept
mse = mean_squared_error(y, y_pred)

print("Slope:", slope)
print("Intercept:", intercept)
print("Correlation coefficient:", r_value)
print("Mean Squared Error:", mse)

# Plot
plt.scatter(x, y, color="blue")
plt.plot(x, y_pred, 'r-', label='Fitted Line')
plt.text(1.5, max(y) - 14000,
         f"y = {slope:.2f}x + {intercept:.2f}\n"
         f"r = {r_value:.2f}\nMSE = {mse:.2f}",
         fontsize=12)
plt.xlabel(x_title)
plt.ylabel(y_title)
plt.title("Regression")
plt.legend()
plt.savefig("regression_plot_python.png")
plt.show()
